### ISSUE TITLE
"text state" updated from in parallel but also API accessible

### ISSUE THREAD

**Issue by annevk** [w3c/edit-context#41]
https://w3c.github.io/edit-context/#update-the-text-edit-context seems to suggest that "text state" can be manipulated from in parallel, but it's also directly accessible through the API (e.g., text state's text). That doesn't work. You have to queue tasks to synchronize such state with the main thread.

Also,

1. Terms like "in parallel" need to reference their definition.
2. Variables should be marked up as such.

**Comment by dandclark** [comment]
I think the "in parallel" here can just be dropped. The details of how the `Text Edit Context` is updated are explicitly deferred to implementations, and will differ across platforms. In practice there might be both sync and async aspects to this, e.g. synchronous updates in a render process that will be asynchronously propagated to a browser process, and then requested at some arbitrary interval by the OS's text input service. There are a couple notes in the spec that mention this, so no need to address it in this algo with something like "in parallel".

Created https://github.com/w3c/edit-context/pull/45.

**Comment by annevk** [comment]
Hmm, I don't think that really works @dandclark. We need to be somewhat clear how state updates are driven as they also relate to events and such. This is not a thing we can entirely hand-wave.

**Comment by dandclark** [comment]
The details under the hood of the [Text Edit Context](https://w3c.github.io/edit-context/#dfn-text-edit-context) are going to be need to be OS-specific to a large extent; it's an abstraction around the wdetails of how the OS and the browser communicate via TSF, NSTextInputClient, etc. I don't think the timings of when different pieces of the text input state are updated via these channels can be made to match across platforms, which is why we hand-wave it here.

I think it's conceptually correct that the state update happens "synchronously" here, so that this info is available immediately to the OS. But the OS itself may consume some of this state asynchronously, in a process apart from the browser's event loop.

**Comment by annevk** [comment]
If it's available synchronously to the OS it cannot be synchronously available to the event loop. There's gonna be IPC which requires task queueing.

**Comment by dandclark** [comment]
Would something like this be appropriate?

"6. [Queue a global task ](https://html.spec.whatwg.org/multipage/webappapis.html#queue-a-global-task) on the [DOM manipulation task source](https://html.spec.whatwg.org/multipage/webappapis.html#dom-manipulation-task-source) to update the [Text Edit Context](https://w3c.github.io/edit-context/#dfn-text-edit-context)'s [text state](https://w3c.github.io/edit-context/#dfn-text-state) to match the values in *editContext*'s [text state](https://w3c.github.io/edit-context/#dfn-text-state)"

**Comment by annevk** [comment]
Queue a global task takes more arguments, but something like that could be a reasonable first step. That still doesn't really define interaction with events and such, which will likely dispatch around that time too.

**Comment by dandclark** [comment]
I've made some changes to the spec that should address this. Now the steps in https://w3c.github.io/edit-context/#update-the-text-edit-context are posted to the task queue as stated in https://w3c.github.io/edit-context/#event-loop-changes (where the proposed change in that section will eventually move to the HTML spec).

@annevk does this approach look reasonable to you?

**Comment by dandclark** [comment]
I believe this has been resolved -- see above [comment](https://github.com/w3c/edit-context/issues/41#issuecomment-1716607906).
